lib/docker-file.ts

```diff
@@ -26,10 +26,45 @@ async function readDockerfileAndAnalyse(
   return analyseDockerfile(contents);
 }
 
+function resolveVariables(
+  text: string,
+  instruction: Instruction,
+  dockerfile: Dockerfile,
+): string {
+  const variables = instruction.getVariables();
+  if (variables) {
+    const resolvedVariables = variables.reduce((resolvedVars, variable) => {
+      const line = variable.getRange().start.line;
+      const name = variable.getName();
+      resolvedVars[name] = dockerfile.resolveVariable(name, line);
+      return resolvedVars;
+    }, {});
+    Object.keys(resolvedVariables).forEach((variable) => {
+      text = text.replace(`\$\{${variable}\}`, resolvedVariables[variable]);
+    });
+  }
+  return text;
+}
+
 async function analyseDockerfile(
   contents: string,
 ): Promise<DockerFileAnalysis | undefined> {
   const dockerfile = DockerfileParser.parse(contents);
+  // collect stages names
+  const stages = dockerfile.getFROMs().reduce((names, from) => {
+    const args = from.getArguments();
+    if (args.length > 2 && args[1].getValue() === "AS") {
+      const fromBaseName = resolveVariables(
+        from.getImage() as string,
+        from,
+        dockerfile,
+      );
+      const fromAsName = args[2].getValue();
+      // support nested referral
```

Comment:
Do we have an example of nested referrals in the fixtures ?

Reply:
The fixture https://github.com/snyk/snyk-docker-plugin/pull/98/files#diff-7f8eb961c7fb84a0b2e8d3e780b06bbeR1
The test https://github.com/snyk/snyk-docker-plugin/pull/98/files#diff-fc4c77fa4619a1c0b7c65906e79c7d99R51